### PR TITLE
feat(endo): sketch out the integration of compartment-mapper and endowmentsToolkit

### PR DESCRIPTION
**Pull request by naugtur**
Noticable conceptual differences to what we discussed:

1. I noticed it's possible to pass an attenuator as an exit module assuming a unique name that's not in the compartment map. We need to think through the consequences.
2. I ended up moving the copy call to the attenuator so that there's one less wrapping for the root compartment (that could be ugly and potentially lossy, make implementing the `'write'` permission harder.

Also, remembered that Endo requires a policy for each package instead of assuming empty. We might need to change that for practical reasons. I started out with that assumption in webpack too but it's not what lavamoat used to do and it's super annoying.

### CODE REVIEW

File: packages/endomoat/src/index.js [hunk 1]
```diff
@@ -18,6 +27,21 @@ export const importHook = async (specifier) => {
   })
 }
 
+export const run = async (entrypointPath, policy) => {
+  const { namespace } = await importLocation(readPower, entrypointPath, {
+    policy: toEndoPolicy(policy),
+    globals: globalThis,
+    importHook,
+    modules: {
+      'USE_ME_AS_DEFAULT_ATTENUATOR_NAME': {
```
Comment (boneskull) [inline packages/endomoat/src/index.js]: what does this mean?

**Comment by boneskull** [comment]
@naugtur rebased onto `endo-refactor`

File: packages/endomoat/src/default-attenuator.js [hunk 2]
```diff
@@ -1,62 +1,56 @@
...
-            if (policyValue === WRITE_POLICY) {
-              spec.set = (value) => {
-                console.log('- set', key)
-                globalOverrides[key] = value
-              }
-            }
-            return [key, spec]
-          })
+      rootCompartmentGlobalThis = packageCompartmentGlobalThis
+      // ^ this is a little dumb, but only a little - assuming importLocation is called only once in parallel. The thing is - even if it isn't, the new one will have a new copy of this module anyway.
```
Comment (boneskull) [inline packages/endomoat/src/default-attenuator.js]: I think I see what you're saying, but:

1. Why not `rootCompartmentGlobalThis =?? packageCompartmentGlobalThis`?
2. Why mention "called _once_ in parallel"? 
3. Why would this branch be executed multiple times _at all_?

File: packages/endomoat/src/index.js [hunk 3]
```diff
@@ -3,19 +3,57 @@ import 'ses'
...
+  const { namespace } = await importLocation(readPower, entrypointPath, {
+    policy: toEndoPolicy(policy),
+    globals: globalThis,
+    importHook,
+    modules: {
+      USE_ME_AS_DEFAULT_ATTENUATOR_NAME: {
+        attenuateGlobals: () => {
+          console.log('lol, this works')
+        },
+      },
```
Comment (boneskull) [inline packages/endomoat/src/index.js]: can we just zap this if we're not gonna use it, because having it there confuses my dumb monkey brain

**Review by boneskull — COMMENTED** [review]
fwiw, if you pull down [this branch](https://github.com/boneskull/endo/tree/boneskull/for-endomoat), run `npm link -w @endo/compartment-mapper ses` from yr `endo` working copy, then go into `packages/endomoat` (important) and run `npm link @endo/compartment-mapper ses` you will get types out of these two packages.

File: packages/endomoat/src/index.js (hunk 3 above)
Comment (naugtur) [reply packages/endomoat/src/index.js]: ok. 
I'm not sure if we're guaranteed not to use it, but I don't mind deleting. It's not like we'll forget it existed now :)

File: packages/endomoat/src/default-attenuator.js (hunk 2 above)
Comment (naugtur) [reply packages/endomoat/src/default-attenuator.js]: this branch would only be executed once per importLocation call, but there's nothing stopping you to all it multiple times in one process. It'd become problematic if we called it multiple times in parallel while using a shared instance of the attenuator.

Comment (boneskull) [reply packages/endomoat/src/default-attenuator.js]: Would it be worthwhile to guard against this?  e.g.,

```js
function setRootCompartmentGlobalThis(obj) {
  rootCompartmentGlobalThis ??= obj
}

// or

function setRootCompartmentGlobalThis(obj) {
  if (rootCompartmentGlobalThis && obj !== rootCompartmentGlobalThis) {
    throw new ReferenceError('ya dun goofed')
  }
  rootCompartmentGlobalThis ??= obj
}

// and in attenuateGlobals
if (policy === ROOT_POLICY) {
  setRootCompartmentGlobalThis(packageCompartmentGlobalThis)
}
```

**Comment by boneskull** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/744?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#744** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/744?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#742** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/742?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#752** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/752?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#761](https://github.com/LavaMoat/LavaMoat/pull/761) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/761?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#770](https://github.com/LavaMoat/LavaMoat/pull/770) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/770?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#732** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/732?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#733](https://github.com/LavaMoat/LavaMoat/pull/733) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/733?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#755](https://github.com/LavaMoat/LavaMoat/pull/755) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/755?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @naugtur and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by boneskull** [comment]
Current dependencies on/for this PR:
* main
  * **PR #732** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/732" target="_blank">Graphite</a> 
    * **PR #752** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/752" target="_blank">Graphite</a> 
      * **PR #753** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/753" target="_blank">Graphite</a> 
    * **PR #742** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/742" target="_blank">Graphite</a> 
      * **PR #744** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/744" target="_blank">Graphite</a>  👈
    * **PR #733** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/733" target="_blank">Graphite</a> 
    * **PR #651** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/651" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/744?utm_source=stack-comment).

**Pull request merged**